### DISCUSSION TITLE
Adding SignUp callback

### DISCUSSION THREAD

**Discussion by zanzlender** [nextauthjs/next-auth#4124] [Ideas]
I have been using NextAuth in a lot of my projects lately and faced some weird challenges while doing so. For example when I have multiple providers, or combined with credentials etc.

Usually what I want to do with my app is something like this: 

LOGIN PAGE
 * user clicks on google login
 * check if this user is in my database (registered)
 * if registered --> signIn
 * if not --> save to db (register), then signIn

It all works great 👍 but it gets kinda messy putting all that logic in, especially because different providers need different logic etc. I don't know if this has been asked before, but is it possible do add a **SignUp** function? It doesn't need to do anything special, it would just be for defining the SignUp callback logic and decoupling it from the SignIn logic.

Then my flow would be something like this: 

LOGIN PAGE
 * user clicks on google login (SignIn)
 * check if this user is in my database 
 * if registered --> signIn
 * if not --> display "User not found"

REGISTRATION PAGE
 * user clicks on register with google (SignUp())
 * check if this user is in my database 
 * if registered --> display "User already exists"
 * if not --> register user, then signIn

**Comment by paulegradie** [comment]
Was this ever implemented? There is a `SignUp` page in the configuration object, but there is no signUp callback. That would be _very_ handy to do thigns such as pre-populate an account with some data when they first create their account. Its particualy important for single-click sign up, e.g. when using one of the database adapters (e.g. the PrismaAdapter).

Definitely a no-brainer feature to live somewhere amongst the auth. 👍

**Comment by KModestas** [comment]
yes please

**Comment by zanzlender** [comment]
This is implemented when creating/using an adapter, it should have a createUser function that is called when a new user tries to log in. [read here](https://next-auth.js.org/configuration/events#createuser)

And also, the pages object has a newUser property where you can redirect him when he registers to do w/e you want. E.g. you could redirect to an api that will redirect to the page you want, or redirect directly to that page. [Auth docs](https://next-auth.js.org/configuration/pages)

However, I couldn't find anywhere in the docs when exactly that callback is fired. E.g. does the adapter immediately query for the user data in db, and if nothing is found automatically calls createUser, or maybe something else... Or rather it does say `User creation in the database is automatic and happens when the user is logging in for the first time with a provider`, however, it doesn't mention how they check for that specifically...

**Comment by JoshPav** [comment]
For anyone looking for a workaround, you can get something without too much effort by wrapping the default methods of your desired adapater. 

Here's a rough example of how I've got a `signUp` callback working with the `MongoDBAdapter`.

```typescript
// /api/auth/[...nextauth]/route.ts

...

const DBAdapter = MongoDBAdapter(clientPromise)

export const authOptions: AuthOptions = {
    /* Your other options here */
    adapter: {
        ...DBAdapter,
        createUser: (async user) => {

            // Using non-null assertion as according to NextAuth this method is requried.  
            const createdUser = await DBAdapter.createUser!(user)
            
            // Run whatever callbacks you want here
            prePopulateUserData(createdUser); 

            return createdUser;
        }
    },
}
```

Of course you can extend this to work for all the other methods used by your `Adapter`.  Hope this is of use to some others!

**Reply by andresgutgon** [reply]
This is super nice. Thanks

**Reply by zgababa** [reply]
Nice, but how do you handle redirect ? I would like to redirect new users on /welcome for example.

**Comment by imyuanli** [comment]
这是自动回复的邮件。您好，我最近正在休假中，无法亲自回复你的邮件。我将在假期结束后，尽快给你回复。

**Reply by andresgutgon** [reply]
yes
